Async coll perm graph revision

escherize opened the pull request:
Pursuant to https://github.com/metabase/metabase/issues/39997

Improves collection graph update performance, especially on problematic instances that have 1000+ collections and perm groups, by making the costly process of writing the potentially gigantic revision json blob to the db into an async process.

An alternative solution is to remove the collection perm graph revision capturing entirely. We still might do that since metabase/metabase does not use the perm graph revision for anything. But it may be nice to have for users and/or support. If it's needed, we could consider a diff-based approach instead.

escherize committed:
Log changes when they make sense + async coll perm-graph revision

escherize commented on src/metabase/models/collection/graph.clj:
```diff
@@ -170,13 +170,15 @@
          new-perms          (into {} (for [[group-id collection-id->perms] new-perms]
                                       [group-id (select-keys collection-id->perms (keys (get old-perms group-id)))]))
          [diff-old changes] (data/diff old-perms new-perms)]
-     (data-perms.graph/log-permissions-changes diff-old changes)
```
We were logging before checking that the revision numbers were correct. That seems wrong, so I change the order that this happens too.

johnswanson reviewed (changes requested):
This makes me a bit nervous - it doesn't provide the same concurrency guarantees as before.

Say `save-perms-revision!` takes 60s to complete. We'll update the permissions graph, but the revision won't be updated for a minute. So e.g. say I am on revision 47 and make a change. The endpoint returns quickly. At this moment, someone else makes a different change to permissions - also with revision 47. `check-revision-numbers` won't throw because as far as it knows we're still on revision 47. It'll successfully change the permissions graph - then sometime in the future we'll get an exception when we try to insert two revisions with the same ID (one will be lost).

In the end we'll have a single new revision (48) and I think it'll have the wrong diff.

One option would be to decouple the revision number from the actual revision write itself (maybe we could insert it and then fill it in later?) but I'm not really sure.

escherize commented:
Good call, I will decouple incrementing the revision number (fast, and should be sync) from writing the perm-graph (slow and will be async).

escherize committed:
- Make revision-number consistent

Decouples the creation of a revision number from the (potentially
expensive) writing of before and after for the coll perms-graph
revision.

escherize commented on src/metabase/models/collection/graph.clj:
```diff
@@ -153,6 +152,22 @@
+    (first (t2/insert-returning-instances! :model/CollectionPermissionGraphRevision
+                                           :id      (inc current-revision-number)
+                                           :user_id api/*current-user-id*
+                                           :before ""
+                                           :after ""))))
```
I almost wrote a migration here, but was swayed by: https://github.com/metabase/metabase/blob/async-coll-perm-graph-revision/resources/migrations/001_update_migrations.yaml#L8267

escherize committed:
get id from the perms revision entity

escherize committed:
Move the revision id creation into the transaction

The revision id creation should only happen iff the perm updates went
through successfully.

escherize committed:
write a test that checks the incomplete revision

johnswanson reviewed (approved):
Nice, LGTM!

noahmoss commented on src/metabase/models/collection/graph.clj:
```diff
@@ -153,6 +152,23 @@
+  [current-revision-number]
+  (when api/*current-user-id*
+    (first (t2/insert-returning-instances! :model/CollectionPermissionGraphRevision
+                                           :id      (inc current-revision-number)
```
nit
```suggestion
                                           :id (inc current-revision-number)
```

noahmoss reviewed (commented):
Filling in the before/after async makes sense to me. For testing, I wonder if we can just deref the future returned from `update-graph!` so that the test blocks until the data is written? Is there anything that requires the wait-until approach? Hard-coded sleeps in tests just make me nervous.

escherize committed:
remove sleep from tests
